### PR TITLE
Add test actions

### PR DESCRIPTION
**Pull request by ivanpauno**
Connected to #176. This pull request add two actions GTest and PyTest, to run tests.
I'm doing this over #167, as running the old test with the legacy API generated a run-time asyncio error ('event loop is closed'), because of bad test interaction.
Also migrated to ament_cmake package, in order to build the dummy cpp test (maybe it exists a better way of doing this).

### CODE REVIEW

File: launch_testing/dummy_tests/locking.py [hunk 1]
```diff
@@ -12,12 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (hidmic) [inline launch_testing/dummy_tests/locking.py]: @ivanpauno please update the copyright year.

File: launch_testing/launch_testing/__init__.py [hunk 2]
```diff
@@ -22,8 +22,13 @@
 from launch.event_handlers import OnProcessIO
 from launch.events import Shutdown
 
+from . import actions
 from .output import create_output_check
 
+__all__ = [
+    'actions',
+]
+
```
Comment (hidmic) [inline launch_testing/launch_testing/__init__.py]: @ivanpauno I'm not sure we want __all__ to behave like this (see [here](http://xion.io/post/code/python-all-wild-imports.html)). Maybe we should add the class defined below and the `output` module as well?

File: launch_testing/launch_testing/actions/gtest.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
```
Comment (hidmic) [inline launch_testing/launch_testing/actions/gtest.py]: @ivanpauno `2018` -> `2019`

File: launch_testing/launch_testing/actions/pytest.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        Constructor.
+
+        timeout argument is passed to :class:`launch_testing.Test`.
+        The other arguments are passed to :class:`launch.ExecuteProcess`, so
+        see the documentation for the class for additional details.
+
+        :param: path the path of the test to be executed.
+        """
+        cmd = [FindExecutable(name='python3'), '-m pytest', path]
+        super().__init__(cmd=cmd, shell=True, **kwargs)
```
Comment (hidmic) [inline launch_testing/launch_testing/actions/pytest.py]: @ivanpauno I _think_ that using `FindExecutable` precludes the need for `shell=True`.

File: launch_testing/launch_testing/actions/test.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        Delegated to :meth:`launch.actions.ExecuteProcess.execute`.
+        """
+        actions = super().execute(context)
+        if not self.__timeout:
+            return actions
+        # Setup a timer to send us a SIGKILL if the test locks
+        sigkill_timer = TimerAction(period=self.__timeout, actions=[
+            EmitEvent(event=SignalProcess(
+                signal_number=signal.SIGKILL,
```
Comment (hidmic) [inline launch_testing/launch_testing/actions/test.py]: @ivanpauno maybe we could send a `SIGINT` instead? It's slightly more polite :)

File: launch_testing/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+find_package(ament_cmake REQUIRED)
+
+ament_export_dependencies(ament_cmake)
+
+ament_python_install_package(${PROJECT_NAME})
+
+add_executable(locking dummy_tests/locking.cpp)
```
Comment (hidmic) [inline launch_testing/CMakeLists.txt]: @ivanpauno consider relocating your dummy applications below `test/`.

File: launch_testing/test/launch_testing/test_gtest.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+from pathlib import Path
+
+from launch import LaunchDescription
+from launch import LaunchService
+
+from launch_testing.actions import GTest
+
+
+def test_gtest():
```
Comment (hidmic) [inline launch_testing/test/launch_testing/test_gtest.py]: @ivanpauno consider adding tests for real `googletest` and `pytest` instances, here and below respectively.

File: launch_testing/launch_testing/actions/test.py (hunk 5 above)
Comment (ivanpauno) [reply launch_testing/launch_testing/actions/test.py]: Sounds good, but maybe it should be escalated to SIGKILL after an extra timeout. WDYT?

Comment (hidmic) [reply launch_testing/launch_testing/actions/test.py]: That's `ExecuteProcess` action behavior upon shutdown already. Maybe we could try to reuse its `__shutdown_process()` method somehow?

**Comment by ivanpauno** [comment]
@wjwwood I'm having a problem here, in a test that is locking forever.
I'm adding two actions to run gtest and pytest, which after a timeout are killed. I'm reusing the on_shutdown  actions of ExecuteProcess, in order to not re-implement the same thing again.
The first timer is ending fine, SIGINT is being emited, and the gtest action is ending fine. Then the two other timers (for escalating to SIGTERM and SIGKILL) are cancelled, but the loop continues waiting for them and never ends. I don't know if this behavior is ok, can you check?

How to repro:
- I added an extra on_process_exit handler to avoid the error in test_gtest.py, comment it:
https://github.com/ros2/launch/blob/cea2cb23b8bacd2f2b6b5529fd3246095118045b/launch_testing/test/launch_testing/test_gtest.py#L41-L43
- It's better only running the test_gtest.py, and not all the tests.

File: launch_testing/launch_testing/actions/pytest.py (hunk 4 above)
Comment (ivanpauno) [reply launch_testing/launch_testing/actions/pytest.py]: It fails without shell=True

**Comment by wjwwood** [comment]
Those timers do need to hold up shutdown, but not all timers need to do so (see: https://github.com/ros2/launch/pull/181), but `ExecuteProcess` action cancels those timers when the process exits before they expire, preventing them from holding up shutdown when the process exits quickly. Perhaps you're missing that part.

I haven't had time to fully review this, so I can't speak to how you're reusing the parts of execute process, though I'm a bit curious why you're not inheriting from it if you're both running a subprocess and wanting it to be killed gracefully (with escalating signals).

**Comment by hidmic** [comment]
@wjwwood the added `Test` action and their subclass do inherit from `ExecuteProcess`. @ivanpauno showed me the issue he's describing above. It's not that timers are holding up the loop, but the `shutdown_on_idle` feature that (it seems to me) has a race. 

- The executable exits and a `ProcessExit` event is sent.
- The timers are canceled (i.e. their cancelling future is resolved).
- The **main loop awakes first** and handles the event sent.
- The main loop goes on and awaits for more events since the timers' completion futures have not been resolved yet.
- The timers' coroutines are awaken and their completion futures resolved.
- The main loop never wakes up because no event was emitted nor will be emitted.

**Comment by hidmic** [comment]
On a side note, may I ask you guys  @mjcarroll @sloretz (and maybe @wjwwood if you feel like it) to review this? I've touched it, so I can't review it anymore.

**Comment by wjwwood** [comment]
All the waiting occurs here:

https://github.com/ros2/launch/blob/d5ace5084262de96115ea7d68b1c43160b0bf3e7/launch/launch/launch_service.py#L230-L237

Do you see debug output for "still waiting on ..."? I'm not initially convinced by your bullet point scenario above that there is a race, but it's been a while since I've steeped myself in this code, so maybe.

**Comment by hidmic** [comment]
But that's if you've received a shutdown event. In the case described above, we were hoping the loop would trigger the shutdown itself once it detects it's idle:

https://github.com/ros2/launch/blob/d5ace5084262de96115ea7d68b1c43160b0bf3e7/launch/launch/launch_service.py#L210-L214

But it seems to be getting stuck in:

https://github.com/ros2/launch/blob/d5ace5084262de96115ea7d68b1c43160b0bf3e7/launch/launch/launch_service.py#L239-L240

**Comment by hidmic** [comment]
@mjcarroll ping

**Comment by hidmic** [comment]
@mjcarroll can we get this one in? I need to the test action concept for https://github.com/ApexAI/apex_rostest/issues/19. 

Running CI just in case:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6391/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2783/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5258/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6230/)

**Comment by hidmic** [comment]
And that's why running CI is important :)

**Comment by hidmic** [comment]
Alright, CI is passing. Merging! Thanks for taking a look @mjcarroll !

**Pull request merged**